### PR TITLE
Improve `torch.masked.mean` and `torch.masked._std_var` scaling

### LINKED ISSUE #131292: Poor scaling of `torch.masked.mean`
### 🐛 Describe the bug

Consider this equivalent formulation (for strided tensors, at least):

```python
def masked_mean(
    input: torch.Tensor,
    mask: torch.Tensor,
    dim: int | torch.types._size | None = None,
    keepdim: bool = False,
    dtype: torch.dtype | None = None,
) -> torch.Tensor:
    return (input * mask).sum(dim=dim, keepdim=keepdim, dtype=dtype or input.dtype) / mask.broadcast_to(input.shape).sum(dim=dim, keepdim=keepdim)
```

21232fda-366a-47ea-a017-15a35cd51d0c

It is due to the `count` calculation:

https://github.com/pytorch/pytorch/blob/e506dfa640639b381efa3ae87b4992d40e8b4d6a/torch/masked/_ops.py#L1388-L1393

Allocating a new tensor is both expensive and unnecessary. Notably, this is also done in `torch.masked._std_var`:

https://github.com/pytorch/pytorch/blob/e506dfa640639b381efa3ae87b4992d40e8b4d6a/torch/masked/_ops.py#L1612-L1617

Just switching this line to:

```python
count = inmask.sum(dim=dim, keepdim=keepdim)
```

You can see the scaling matches the proposed implementation:

aad2dfcc-52c9-4046-86ab-122b044fa19c

Here is the code I used to generate the plots:

```python
import time

import matplotlib.pyplot as plt
import torch
import torch.types

num_elements: list[int] = []
avg_times_custom: list[float] = []
avg_times_torch_masked_mean: list[float] = []

# takes about 3 minutes on my laptop
NUM_STEPS = 16
RUNS_PER_STEP = 5
MAX_NUM_ELEMENTS = 4096 * 4096 * 32


def masked_mean(
    input: torch.Tensor,
    mask: torch.Tensor,
    dim: int | torch.types._size | None = None,
    keepdim: bool = False,
    dtype: torch.dtype | None = None,
) -> torch.Tensor:
    return (input * mask).sum(dim=dim, keepdim=keepdim, dtype=dtype or input.dtype) / mask.broadcast_to(input.shape).sum(dim=dim, keepdim=keepdim)


for n_elements in torch.linspace(1, MAX_NUM_ELEMENTS, NUM_STEPS):
    n_elements = int(n_elements)
    num_elements.append(n_elements)

    input_tensor = torch.randn(n_elements, dtype=torch.float32)
    mask = torch.tensor(True)
    # mask = torch.randint_like(input_tensor, 0, 2, dtype=torch.bool)  # non-trivial mask

    torch.testing.assert_close(masked_mean(input_tensor, mask=mask), torch.masked.mean(input_tensor, mask=mask))  # will fail if not mask.any()

    total_time_custom = 0.0

    for _ in range(RUNS_PER_STEP):
        start_time = time.perf_counter()
        masked_mean(input_tensor, mask=mask)
        end_time = time.perf_counter()
        total_time_custom += (end_time - start_time)

    avg_time_cusotm = (total_time_custom / RUNS_PER_STEP)
    avg_times_custom.append(avg_time_cusotm)

    total_time_torch_masked_mean = 0.0

    for _ in range(RUNS_PER_STEP):
        start_time = time.perf_counter()
        torch.masked.mean(input_tensor, mask=mask)
        end_time = time.perf_counter()
        total_time_torch_masked_mean += (end_time - start_time)

    avg_time_torch_masked_mean = (total_time_torch_masked_mean / RUNS_PER_STEP)
    avg_times_torch_masked_mean.append(avg_time_torch_masked_mean)

plt.figure(figsize=(10, 6))
plt.plot(num_elements, avg_times_custom, label='proposed implementation (without new_ones)')
plt.plot(num_elements, avg_times_torch_masked_mean, label='torch.masked.mean')
plt.scatter(num_elements, avg_times_custom)
plt.scatter(num_elements, avg_times_torch_masked_mean)
plt.xlabel('Number of Elements')
plt.ylabel('Average Execution Time (s)')
plt.title('Average Execution Time Scaling')
plt.legend()
plt.show()
```

### Versions

torch v2.3.1

### PR DESCRIPTION
**Pull request by ringohoffman**
Fixes #131292

Using `new_ones` is expensive and unnecessary.

Before:

21232fda-366a-47ea-a017-15a35cd51d0c

After:

aad2dfcc-52c9-4046-86ab-122b044fa19c

cc: @ezyang @george-qi

### CODE REVIEW

File: torch/masked/_ops.py [hunk 1]
```diff
@@ -1385,12 +1385,7 @@ def mean(
             total = sum(input, dim, keepdim=keepdim, dtype=dtype)
         else:
             inmask = _input_mask(input, mask=mask)
-            count = sum(
-                inmask.new_ones(input.shape, dtype=torch.int64),
-                dim,
-                keepdim=keepdim,
-                mask=inmask,
-            )
+            count = inmask.sum(dim=dim, keepdim=keepdim)
```
Comment (vadimkantorov) [inline torch/masked/_ops.py:1388]: If `inmask` is bool, doing sum currently would also upcast/allocate a not very required temp copy I think:
- https://github.com/pytorch/pytorch/issues/55366

Comment (ringohoffman) [reply torch/masked/_ops.py:1388]: I think it is guaranteed to be bool. Sorry if I'm not understanding, but did you have a suggested change to make?

Comment (vadimkantorov) [reply torch/masked/_ops.py:1388]: I'm not a maintained, so my personal opinion is: it's already nice that the PR fixes the scaling problem, but maybe this issue that `bool_tensor.sum()` allocates a temp/upcast copy of the input in eager mode could be bumped

**Review by ezyang — APPROVED** [review]
